php/src/Google/Protobuf/Internal/Message.php

```diff
@@ -786,27 +827,99 @@ private function convertJsonValueToProtoValue(
                 }
-            case GPBType::STRING:
+                $proto_value = base64_decode($value, true);
+                if ($proto_value === false) {
+                    throw new GPBDecodeException(
+                        "Invalid base64 characters");
+                }
+                return $proto_value;
+            case GPBType::BOOL:
                 if (is_null($value)) {
                     return $this->defaultValue($field);
                 }
-                if (!is_string($value)) {
-                    throw new GPBDecodeException("Expect string");
+                if ($is_map_key) {
+                    if ($value === "true") {
+                        return true;
+                    }
+                    if ($value === "false") {
+                        return false;
+                    }
+                    throw new GPBDecodeException(
+                        "Bool field only accept bool value");
+                }
+                if (!is_bool($value)) {
+                    throw new GPBDecodeException(
+                        "Bool field only accept bool value");
+                }
+                return $value;
+            case GPBType::INT64:
+            case GPBType::SINT64:
+            case GPBType::SFIXED64:
+                if (is_null($value)) {
+                    return $this->defaultValue($field);
+                }
+                if (!is_numeric($value)) {
+                   throw new GPBDecodeException(
+                       "Invalid data type for int64 field");
+                }
+                if (bccomp($value, "9223372036854775807") > 0) {
+                    throw new GPBDecodeException(
+                        "Int64 too large");
+                }
+                if (bccomp($value, "-9223372036854775808") < 0) {
+                    throw new GPBDecodeException(
+                        "Int64 too small");
+                }
+                return $value;
+            case GPBType::UINT64:
+            case GPBType::FIXED64:
+                if (is_null($value)) {
+                    return $this->defaultValue($field);
+                }
+                if (!is_numeric($value)) {
+                   throw new GPBDecodeException(
+                       "Invalid data type for int64 field");
+                }
+                if (bccomp($value, "18446744073709551615") > 0) {
+                    throw new GPBDecodeException(
+                        "Uint64 too large");
+                }
+                if (bccomp($value, "9223372036854775807") > 0) {
+                    $value = bcsub($value, "18446744073709551616");
+                }
+                return $value;
+        }
+        return $this->convertNativeValueToProtoValue(
+            $value,
+            $field,
+            $is_map_key);
+    }
+
+    private function convertNativeValueToProtoValue(
+        $value,
+        $field,
+        $is_map_key = false)
+    {
+        if (is_null($value)) {
```

Comment:
For json, null is acceptable for primitive fields. However, for array constructor, null is not acceptable for primitive fields. You need to add check for this case.

Reply:
What should happen in the array constructor case? Do you want an exception thrown? e.g:

```php
new TestMessage(['intField' => null]);

// @throws InvalidArgumentException "field intField is an int and cannot be null"
```

If we do this, however, it's inconsistent, as it is possible to call `$testMessage->setIntField(null);`

Reply:
Nevermind! I'm wrong. I will follow the existing behavior, which is the following:

```php
if (is_numeric($var)) {
    $var = intval($var);
} else {
    throw new \Exception("Expect integer.");
}
```

Reply:
Please follow the ```checkInt32``` in GPBUtil for this case.